src/rng.jl

```diff
@@ -0,0 +1,104 @@
+using Random123
+using Random
+using Distributions
+
+import Base.rand
+import Random.seed!
+import Random123: set_counter!
+
+# Default RNG type for when nothing is specified
+_BASE_RNG = Philox2x
+
+"""
+    TracedRNG{R,T}
+
+Wrapped random number generator from Random123 to keep track of random streams during model evaluation
+"""
+mutable struct TracedRNG{T} <:
+               Random.AbstractRNG where {T<:(Random123.AbstractR123{R} where {R})}
+    "Model step counter"
+    count::Int
+    "Inner RNG"
+    rng::T
+    "Array of keys"
+    keys
```

Comment:
Make this concretely typed as well?

Reply:
Maybe something like this: 
```julia
mutable struct TracedRNG{R, T<:Random123.AbstractR123{R}} <: Random.AbstractRNG
    "Model step counter"
    count::Int
    "Inner RNG"
    rng::T
    "Array of keys"
    keys::Array{R}
end
```

Reply:
`Array{R}` is not a concrete type but e.g. `Vector{R}`, `Matrix{R}` or `Array{R,N}` (with additional type parameter `N`) would be.

Reply:
My bad, we need `Array{R,N}`. I'll change the types accordingly.